File: packages/events/src/CloudEvent.ts
import { v4 as uuidv4 } from 'uuid';
import { CloudEvent } from './types';

export class CloudEventBuilder {
  private event: CloudEvent;

  constructor(type: string, source: string) {
    this.event = {
      specversion: '1.0',
      type,
      source,
      id: uuidv4(),
      time: new Date().toISOString(),
      datacontenttype: 'application/json',
      attributes: {},
    };
  }

  withId(id: string): CloudEventBuilder {
    this.event.id = id;
    return this;
  }

  withSubject(subject: string): CloudEventBuilder {
    this.event.subject = subject;
    return this;
  }

  withDataSchema(schema: string): CloudEventBuilder {
    this.event.dataschema = schema;
    return this;
  }

  withData(data: Record<string, any>): CloudEventBuilder {
    this.event.data = data;
    return this;
  }

  withAttribute(key: string, value: string): CloudEventBuilder {
    if (!this.event.attributes) {
      this.event.attributes = {};
    }
    this.event.attributes[key] = value;
    return this;
  }

  build(): CloudEvent {
    this.validate();
    return this.event;
  }

  private validate(): void {
    if (this.event.specversion !== '1.0') {
      throw new Error(`Invalid specversion: ${this.event.specversion}`);
    }
    if (!this.event.type) {
      throw new Error('type is required');
    }
    if (!this.event.source) {
      throw new Error('source is required');
    }
    if (!this.event.id) {
      throw new Error('id is required');
    }
  }
}

export class CloudEventValidator {
  static validate(event: CloudEvent): { valid: boolean; errors: string[] } {
    const errors: string[] = [];

    if (event.specversion !== '1.0') {
      errors.push(`Invalid specversion: ${event.specversion}`);
    }
    if (!event.type) {
      errors.push('type is required');
    }
    if (!event.source) {
      errors.push('source is required');
    }
    if (!event.id) {
      errors.push('id is required');
    }
    if (!event.time) {
      errors.push('time is required');
    }
    if (!event.datacontenttype) {
      errors.push('datacontenttype is required');
    }

    return {
      valid: errors.length === 0,
      errors,
    };
  }
}

File: services/event-bus/internal/events/cloudevent.go
package events

import (
	"encoding/json"
	"fmt"
	"time"
)

type CloudEvent struct {
	SpecVersion     string                 `json:"specversion"`
	Type            string                 `json:"type"`
	Source          string                 `json:"source"`
	ID              string                 `json:"id"`
	Time            time.Time              `json:"time"`
	DataContentType string                 `json:"datacontenttype"`
	Subject         string                 `json:"subject,omitempty"`
	DataSchema      string                 `json:"dataschema,omitempty"`
	Data            map[string]interface{} `json:"data,omitempty"`
	Attributes      map[string]string      `json:"attributes,omitempty"`
}

func NewCloudEvent(eventType string, source string, id string, data map[string]interface{}) *CloudEvent {
	return &CloudEvent{
		SpecVersion:     "1.0",
		Type:            eventType,
		Source:          source,
		ID:              id,
		Time:            time.Now().UTC(),
		DataContentType: "application/json",
		Data:            data,
		Attributes:      make(map[string]string),
	}
}

func (ce *CloudEvent) Validate() error {
	if ce.SpecVersion != "1.0" {
		return fmt.Errorf("invalid specversion: %s", ce.SpecVersion)
	}
	if ce.Type == "" {
		return fmt.Errorf("type is required")
	}
	if ce.Source == "" {
		return fmt.Errorf("source is required")
	}
	if ce.ID == "" {
		return fmt.Errorf("id is required")
	}
	return nil
}

func (ce *CloudEvent) MarshalJSON() ([]byte, error) {
	type Alias CloudEvent
	return json.Marshal(&struct {
		Time string `json:"time"`
		*Alias
	}{
		Time:  ce.Time.Format(time.RFC3339Nano),
		Alias: (*Alias)(ce),
	})
}

func (ce *CloudEvent) UnmarshalJSON(data []byte) error {
	type Alias CloudEvent
	aux := &struct {
		Time string `json:"time"`
		*Alias
	}{
		Alias: (*Alias)(ce),
	}

	if err := json.Unmarshal(data, &aux); err != nil {
		return err
	}

	t, err := time.Parse(time.RFC3339Nano, aux.Time)
	if err != nil {
		return err
	}

	ce.Time = t
	return nil
}

func (ce *CloudEvent) ToMap() map[string]interface{} {
	return map[string]interface{}{
		"specversion":       ce.SpecVersion,
		"type":              ce.Type,
		"source":            ce.Source,
		"id":                ce.ID,
		"time":              ce.Time.Format(time.RFC3339Nano),
		"datacontenttype":   ce.DataContentType,
		"subject":           ce.Subject,
		"dataschema":        ce.DataSchema,
		"data":              ce.Data,
		"attributes":        ce.Attributes,
	}
}
